### DISCUSSION TITLE
Some way to put all operations inside a transaction?

### DISCUSSION THREAD

**Discussion by revmischa** [prisma/prisma#22493] [Q&A]
### Question

(For more context please see discord: https://discord.com/channels/937751382725886062/1184188075463688392)

So my understanding is that if I start an interactive transaction, I provide a callback which receives a new prisma client instance that is scoped to that transaction.

```ts
async function doStuff() {
  await prisma.$transaction(async (txn) => {
    await stuff1(txn)
    await stuff2(txn)
  })
}
```

My issue is that `stuff1()` and `stuff2()` call lots of other functions, which call other functions, and so on. It is not feasible to pass txn down into every single function call. 
Is there some way to tell prisma that ALL prisma calls inside of this closure should use the current transaction? that is how I assumed it worked at first, but no, you need to use the txn instance. this caused much unpleasantness. 
Something like how django does atomic transactions would be ideal: https://docs.djangoproject.com/en/4.2/topics/db/transactions/

I just would like to start a transaction and have all calls made inside that closure be inside the transaction.

Also, I can't safely tell my code that it may get a transaction instance or a normal client instance
I tried creating:
`type PrismaTxnClient = Omit<PrismaClient, runtime.ITXClientDenyList>`
(not documented anywhere as far as I could tell)

But if I let my functions accept a `PrismaClient | PrismaTxnClient` then it tells me I can't call stuff like `prisma.$transaction()` because it's not defined for `PrismaTxnClient`.
Why not? couldn't it be a no-op if I'm already inside a transaction?

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by revmischa** [comment]
Perhaps a clean solution for this in the future will be [AsyncContext](https://github.com/tc39/proposal-async-context). Can safely store a consistent transaction handle across the entire transaction async execution.

**Reply by Yamanlk** [reply]
I created a package for that which uses [async_hooks](https://nodejs.org/api/async_hooks.html) take a look if you like 
https://transactional.gitbook.io/documentation/clients/prisma

The package is available in npm and the code for the package is fairly simple and available in GitHub https://github.com/Yamanlk/transactional-prisma/tree/main

however i still think there is a lot of room for improvement so any contribution is welcomed :)

**Reply by revmischa** [reply]
Oh very cool!

**Reply by ludralph** [reply]
Hi @revmischa, Did the comments above help resolve your issue?

**Reply by revmischa** [reply]
My feature request still stands

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
